flycheck.el

```diff
@@ -2725,54 +2770,95 @@ current syntax check."
   (flycheck-error-list-refresh)
   (flycheck-hide-error-buffer))
 
+(defun flycheck--empty-variables ()
+  "Empty variables used by Flycheck."
+  (kill-local-variable 'flycheck--idle-trigger-timer)
+  (kill-local-variable 'flycheck--idle-trigger-conditions))
+
 (defun flycheck-teardown ()
-  "Teardown Flycheck in the current buffer..
+  "Teardown Flycheck in the current buffer.
 
 Completely clear the whole Flycheck state.  Remove overlays, kill
 running checks, and empty all variables used by Flycheck."
   (flycheck-safe-delete-temporaries)
   (flycheck-stop)
   (flycheck-clean-deferred-check)
   (flycheck-clear)
-  (flycheck-cancel-error-display-error-at-point-timer))
+  (flycheck-cancel-error-display-error-at-point-timer)
+  (flycheck--clear-idle-trigger-timer)
+  (flycheck--empty-variables))
 
 
 ;;; Automatic syntax checking in a buffer
-(defun flycheck-may-check-automatically (&optional condition)
-  "Determine whether the buffer may be checked under CONDITION.
+(defun flycheck-may-check-automatically (&rest conditions)
+  "Determine whether the buffer may be checked under one of CONDITIONS.
 
 Read-only buffers may never be checked automatically.
 
-If CONDITION is non-nil, determine whether syntax may checked
-automatically according to
+If CONDITIONS are given, determine whether syntax may be checked
+under at least one of them, according to
 `flycheck-check-syntax-automatically'."
   (and (not (or buffer-read-only (flycheck-ephemeral-buffer-p)))
        (file-exists-p default-directory)
-       (or (not condition)
-           (memq condition flycheck-check-syntax-automatically))))
+       (or (not conditions)
+           (seq-some
+            (lambda (condition)
+              (memq condition flycheck-check-syntax-automatically))
+            conditions))))
+
+(defvar-local flycheck--idle-trigger-timer nil
+  "Timer used to trigger a syntax check after an idle delay.")
+
+(defvar-local flycheck--idle-trigger-conditions nil
+  "List of conditions under which an idle syntax check will be triggered.
+This will be some subset of the allowable values for
+`flycheck-check-syntax-automatically'.
+
+For example, if the user switches to a buffer and then makes an
+edit, this list will have the values `idle-change' and
+`idle-buffer-switch' in it, at least until the idle timer
+expires.")
 
 (defun flycheck-buffer-automatically (&optional condition force-deferred)
   "Automatically check syntax at CONDITION.
 
 Syntax is not checked if `flycheck-may-check-automatically'
-returns nil for CONDITION.
+returns nil for CONDITION.  (CONDITION may be a single condition
+or a list of them.)
```

Comment:
Wouldn't it be simpler to force `condition` to always be a list? Does any external code depend on accepting non-lists?

Reply:
`-` vs `--`: mostly historical reasons, I think; this is a pretty recent convention AFAIK.
Most of the uses I could find are from the same snippet that we discussed above.  @fmdkdd , WDYT?

Reply:
> - vs --: mostly historical reasons, I think; this is a pretty recent convention AFAIK.

I'm not against starting using it for new code (as in this PR).  My main contention is that it's just that: a convention.  We have no way to refrain people from monkey-patching our "private API" functions.  We can just signal that "you are on your own if we remove it in the future".

Regarding the `condition` parameter: accepting single conditions and lists of condition is fine with me.  The code is not that complex.

Reply:
Single conditions and lists of conditions are fine from a maintainability perspective, I think. However, what _is_ a bit confusing is that the `condition` parameter is optional, so passing an "empty list" is the same as omitting the argument (which has a different effect than passing an empty list would be expected to have). How about making the argument non-optional, and adding a special value (some symbol) that can be passed to recover the current behavior for the optional argument being omitted? (This would allow me to remove at least one other place in the code where I intentionally refrain from calling `flycheck-buffer-automatically` when the condition list is empty.)